### PR TITLE
OrganizeImports: optionally move `given` imports to end

### LINKED ISSUE #2255: Organizing imports with ...{given Decoder[X]} fails
Given the following imports:

```
import io.circe.Json
import io.circe.parser.*
import io.circe.syntax.*
import io.circe.{Decoder, Encoder}
import org.scalatest.flatspec.AnyFlatSpec
import chimp.mcp.JSONRPCMessage.{given Decoder[JSONRPCMessage], given Encoder[JSONRPCMessage]}
import org.scalatest.matchers.should.Matchers
```

Organizing them yields:

```
import chimp.mcp.JSONRPCMessage.given Decoder[JSONRPCMessage]
import chimp.mcp.JSONRPCMessage.given Encoder[JSONRPCMessage]
import io.circe.Decoder
import io.circe.Encoder
import io.circe.Json
import io.circe.parser.*
import io.circe.syntax.*
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should.Matchers
```

Which is incorrect, since Decoder is imported after the given.

This has been observed in Metals 1.5.3

### PR DESCRIPTION
**Pull request by kitbellew**
Fixes #2255.

### CODE REVIEW

**Comment by kitbellew** [comment]
@bjaglin ptal

File: scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*
+rules = [OrganizeImports]
+OrganizeImports {
+  groupedImports = Keep
+  removeUnused = false
+  targetDialect = Scala3
```
Comment (bjaglin) [inline scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala]: nit: that is implicit since the harness (whether the testkit here or sbt-scalafix in real life) injects the scala version which is then used to infer the dialect
```suggestion
```

File: scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+}
+ */
+package test.organizeImports
+
+import GivensGeneric.AA
+import test.organizeImports.Givens.A
+import test.organizeImports.Givens.{given AA[A], given A}
+
+object RelativeGenericGivens
```
Comment (bjaglin) [inline scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala:16]: although tricky and thus good to test, the "relative" part seems to be a corner case - the reported issue was with fully qualified references, so I think we should have another test where the given would end up _before_ the generic it references (because of import order) without the fix

File: scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/*
+rules = [OrganizeImports]
+OrganizeImports {
+  groupedImports = Keep
```
Comment (bjaglin) [inline scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala]: is this needed to reproduce the problem? it looks orthogonal
```suggestion
```

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala [hunk 4]
```diff
@@ -225,11 +225,14 @@ class OrganizeImports(
...
     val noImplicitImporters = Seq.newBuilder[Importer]
-    val separateImplicits = config.groupExplicitlyImportedImplicitsSeparately
+    val separateImplicits = config.groupExplicitlyImportedImplicitsSeparately ||
+      targetDialect.allowGivenUsing
     if (separateImplicits) importers.foreach { importer =>
       val (implicits, noImplicits) = importer.importees.partition {
         case i: Importee.Name =>
+          config.groupExplicitlyImportedImplicitsSeparately &&
           i.symbol.infoNoThrow.exists(_.isImplicit)
+        case i: Importee.Given => i.tpe.is[Type.Apply]
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala]: can you document this new scenario in https://github.com/scalacenter/scalafix/blob/main/docs/rules/OrganizeImports.md?plain=1#L757-L758 ?

**Review by bjaglin — COMMENTED** [review]
Thanks! Implementation LGTM, a few comments on docs & testing.

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala (hunk 4 above)
Comment (kitbellew) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala]: will do

File: scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala (hunk 3 above)
Comment (kitbellew) [reply scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala]: the default for `groupedImports` is `Explode`. i used `Keep` (or could have used `Merge`) to request that the grouped imports below on line 13 should be preserved, but they will nevertheless be separated since there's an overriding reason.

i can remove it as the difference will be a blank line between the groups, but this looked a bit more emphatic (and in fact i might even use a more aggressive setting, just to be clear).

File: scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala (hunk 1 above)
Comment (kitbellew) [reply scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala]: done

File: scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala (hunk 2 above)
Comment (kitbellew) [reply scalafix-tests/input/src/main/scala-3/test/organizeImports/RelativeGenericGivens.scala:16]: by "relative" i mean that `AA` used in `given AA[A]` is relative (instead of fully-qualified).

the "trick" with using relative `GivensGeneric.AA` was so that it's "sorted" after the given. let me vary the cases a bit.

**Comment by kitbellew** [comment]
i think this is the wrong approach :( `given` can import absolutely any type, even if it's not `A[B]`, so correct logic needs to figure out the right fully-qualified type of what it is importing. `.symbol` on these imported types returns empty; does that mean semanticdb is incomplete in scala3?

**Comment by bjaglin** [comment]
> does that mean semanticdb is incomplete in scala3?

Yes - https://github.com/scalacenter/scalafix/issues/2054

> i think this is the wrong approach :( given can import absolutely any type, even if it's not A[B], so correct logic needs to figure out the right fully-qualified type of what it is importing.

From my understanding, the current logic in this PR might bring false positives (imports sorted last for "nothing"), but it's still better than generating invalid code, no?

**Comment by albertikm** [comment]
> > does that mean semanticdb is incomplete in scala3?
> 
> Yes - #2054
> 
> > i think this is the wrong approach :( given can import absolutely any type, even if it's not A[B], so correct logic needs to figure out the right fully-qualified type of what it is importing.
> 
> From my understanding, the current logic in this PR might bring false positives (imports sorted last for "nothing"), but it's still better than generating invalid code, no?

the problem is that in reality, short of putting all givens in that separate section, we can't really fix anything that is likely to occur.

the given declaration that we're importing can be of any type (including simple Type.Name), and that type can be coming from any package(s); I thought I could access `.symbol` on all those types, to figure out where to sort the given, but symbols are empty for some reason.

perhaps you could briefly explain how symbols or occurrences can work, and what the current limitations are? I'm also not sure where scala3 semantic information is coming from, and whether it is compatible or vastly different from the scalameta version.

i can send a patch of what i experimented with, to see if it can be improved somehow.

**Comment by kitbellew** [comment]
oops... again used other profile above.

**Comment by kitbellew** [comment]
> > does that mean semanticdb is incomplete in scala3?
> 
> Yes - #2054

Oy vey. And imports are not mentioned there, perhaps they are covered by one of them.

> From my understanding, the current logic in this PR might bring false positives (imports sorted last for "nothing"), but it's still better than generating invalid code, no?

In that case, I would suggest the following: if `groupExplicitlyImportedImplicitsSeparately` is set, move all givens to the end, together with implicits. (Or we could *always* do it.)

But I don't like the logic I have proposed so far, it solves one particular rare case, but doesn't solve other problematic cases which are, in my opinion, even more likely.

**Comment by kitbellew** [comment]
added another option to `groupSeparately` parameter (which was renamed), and moved all givens if the new option is set. i don't see a better solution until semanticdb can help.

File: docs/rules/OrganizeImports.md [hunk 5]
```diff
@@ -437,12 +450,23 @@ details).
...
+
+If this value is specified, any given imports will be put in that
+final catch-all group.
+
+The reason is that a `given` import refers to a type `T`, and that type can
+come from any package `P` and must be imported earlier
+- however, our grouping and sorting logic could put `import P.T` later
+- at the moment, scala 3 semanticdb doesn't provide symbol information for `T`
+
 ### Value type
```
Comment (bjaglin) [inline docs/rules/OrganizeImports.md]: this and the lines below look obsolete - I think we should follow the structure of https://github.com/scalacenter/scalafix/blob/63be364161e828ac5b2bf0e32d9a52255811a2a1/docs/rules/OrganizeImports.md?plain=1#L777-L813 to document that new flag that can take an array (even though we obviously only have a limited set of possible values here)

Comment (bjaglin) [reply docs/rules/OrganizeImports.md]: Also, I think the new feature deserves a new example

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala [hunk 6]
```diff
@@ -227,11 +227,15 @@ class OrganizeImports(
...
+    val separateImplicits = config.groupSeparately
+      .contains(GroupSeparately.ExplicitlyImportedImplicits)
+    val separateGivens = targetDialect.allowGivenUsing &&
+      config.groupSeparately.contains(GroupSeparately.Givens)
+    if (separateImplicits || separateGivens) importers.foreach { importer =>
       val (implicits, noImplicits) = importer.importees.partition {
         case i: Importee.Name =>
-          i.symbol.infoNoThrow.exists(_.isImplicit)
+          separateImplicits && i.symbol.infoNoThrow.exists(_.isImplicit)
+        case _: Importee.Given | _: Importee.GivenAll => separateGivens
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala]: Do we have any problem to address with `GivenAll`? My understanding is that we don't have any information about Y in `import X.given Y`, but we do have information about X to sort `import X.given` properly.
```suggestion
        case _: Importee.Given => separateGivens
```
In any case, we are lacking a test for `GivenAll`.

File: docs/rules/OrganizeImports.md [hunk 7]
```diff
@@ -437,12 +450,23 @@ details).
...
 > minimize manual intervention and guarantee correctness in all cases.
 
-> The `groupExplicitlyImportedImplicitsSeparately` option has currently no
+> This option has currently no
 > effect on source files compiled with Scala 3, as the [compiler does not
 > expose full signature
 > information](https://github.com/scala/scala3/issues/12766), preventing
 > the rule to identify imported implicits.
 
+`groupSeparately`: `Givens`
```
Comment (bjaglin) [inline docs/rules/OrganizeImports.md]: - I think we can clarify that this only impacts [by-type givens](https://docs.scala-lang.org/scala3/reference/contextual/given-imports.html#importing-by-type) (and document the preserved behavior of by-name givens with tests)
- Even though it's obvious, this doc is being very verbose about what's Scala 3 or not, so we might just add a note
```suggestion
`groupSeparately`: ` ByTypeGivens` (Scala 3)
```

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImportsConfig.scala [hunk 8]
```diff
@@ -58,11 +58,20 @@ object TargetDialect {
...
+  implicit val codec: ConfCodecEx[GroupSeparately] = OrganizeImportsConfig
+    .getCodecFrom(ExplicitlyImportedImplicits, Givens)
+}
+
 final case class OrganizeImportsConfig(
     blankLines: BlankLines = BlankLines.Auto,
     coalesceToWildcardImportThreshold: Option[Int] = None,
     expandRelative: Boolean = false,
-    groupExplicitlyImportedImplicitsSeparately: Boolean = false,
+    groupSeparately: Seq[GroupSeparately] = Nil,
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImportsConfig.scala]: I think we should make this the default so we guarantee correctness for whoever uses by-type givens. This is a breaking change for the lucky ones that have been using scalafix and had the proper ordering by chance without a special group but I think it's a tiny price to pay.
```suggestion
    groupSeparately: Seq[GroupSeparately] = Seq(GroupSeparately.ByTypeGivens),
```

File: docs/rules/OrganizeImports.md [hunk 9]
```diff
@@ -437,12 +450,23 @@ details).
...
+--------------------------------------------
+
+If this value is specified, any given imports will be put in that
+final catch-all group.
+
+The reason is that a `given` import refers to a type `T`, and that type can
+come from any package `P` and must be imported earlier
+- however, our grouping and sorting logic could put `import P.T` later
+- at the moment, scala 3 semanticdb doesn't provide symbol information for `T`
```
Comment (bjaglin) [inline docs/rules/OrganizeImports.md]: I went through https://github.com/scalacenter/scalafix/issues/2054 & the Scala 3 upstream issues, and I couldn't even find a ticket tracking that gap. I think it's worth creating one and linking it from the docs.

**Review by bjaglin — COMMENTED** [review]
The metaconfig foo is pretty cool! And the implementation is now very clear about what's going on with that special group.

Yet, if we confirm the fix is only needed for the niche use-case of by-type givens (i.e. not by-names nor wildcard ones), and make it the default for correctness, I wonder if there is really a need for a flag to opt-out?

File: docs/rules/OrganizeImports.md (hunk 7 above)
Comment (kitbellew) [reply docs/rules/OrganizeImports.md]: what are by-name givens?

File: docs/rules/OrganizeImports.md (hunk 9 above)
Comment (kitbellew) [reply docs/rules/OrganizeImports.md]: submitted scala/scala3#26255, could you please add it to #2054?

File: docs/rules/OrganizeImports.md (hunk 5 above)
Comment (kitbellew) [reply docs/rules/OrganizeImports.md]: not sure if i fully understood what you meant, made a small change (mentioned the default and used `####` prefix on values).

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala (hunk 6 above)
Comment (kitbellew) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala]: GivenAll is not needed, perhaps; must have confused with wildcard import which cannot be separated from Unimport/Rename. On the other hand, if we have GivenAll, why would we need explicit givens?

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImportsConfig.scala (hunk 8 above)
Comment (kitbellew) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImportsConfig.scala]: you da boss :)

Comment (kitbellew) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImportsConfig.scala]: this default exposed a possibly overlooked earlier problem, at least one which would otherwise result in massive changes to existing tests. see #2456, need to merge it first, before rebasing this one.

File: scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala (hunk 6 above)
Comment (bjaglin) [reply scalafix-rules/src/main/scala/scalafix/internal/rule/OrganizeImports.scala]: > On the other hand, if we have GivenAll, why would we need explicit givens?

for a given package, we don't - I was just pointing out that GivenAll don't need special treatment since they don't have a RHS where information is missing.

File: docs/rules/OrganizeImports.md (hunk 7 above)
Comment (bjaglin) [reply docs/rules/OrganizeImports.md]: looking at https://docs.scala-lang.org/scala3/reference/contextual/given-imports.html, it's a regular import, referencing a member with the given keyword

Comment (bjaglin) [reply docs/rules/OrganizeImports.md]: actually, it's precisely what ExplicitlyImportedImplicit is for, although in Scala 3 we can't distinguish them

**Comment by kitbellew** [comment]
> ... I wonder if there is really a need for a flag to opt-out?

if we ever get the fix for semanticdb occurrences, we'll be able to sort those givens (and move to catch-all group only those whose dependency is not sorted before them). i don't know if it's ok just to add that new sorting logic and shuffle imports automatically, without a user being able to say "i am ok with them unsorted as before".

but let me know if you think it's ok to omit this. scalafmt used to be very opinionated, but now the opinionated ones are the users :)

File: docs/rules/OrganizeImports.md (hunk 5 above)
Comment (bjaglin) [reply docs/rules/OrganizeImports.md]: sorry, I wasn't clear - I just added a7d6810f27fa81618fc2c04e8584b93cb2995489 to try to improve consistency with the rest of the doc

File: docs/rules/OrganizeImports.md (hunk 7 above)
Comment (bjaglin) [reply docs/rules/OrganizeImports.md]: now that makes me wonder if `ExplicitlyImportedImplicits` shouldn't be called `ByNameImplicits` for symmetry with `ByTypeGivens`

Comment (kitbellew) [reply docs/rules/OrganizeImports.md]: actually, a very sane thought. very much aligned with my brand of OCD :)

**Comment by bjaglin** [comment]
> if we ever get the fix for semanticdb occurrences, we'll be able to sort those givens (and move to catch-all group only those whose dependency is not sorted before them). i don't know if it's ok just to add that new sorting logic and shuffle imports automatically, without a user being able to say "i am ok with them unsorted as before".

Good point. Let's keep it 👍

File: docs/rules/OrganizeImports.md (hunk 7 above)
Comment (albertikm) [reply docs/rules/OrganizeImports.md]: did you want me to rename this?

Comment (bjaglin) [reply docs/rules/OrganizeImports.md]: i think it would make understanding easier, I can do it before merging if you prefer

Comment (kitbellew) [reply docs/rules/OrganizeImports.md]: give me a minute to send in a new revision

Comment (kitbellew) [reply docs/rules/OrganizeImports.md]: all done

**Review by bjaglin — APPROVED** [review]
Thanks!
